File: pythonVersion/market.py
import random

class Market():
    def __init__(self):
        self._companies = []

    def add_company(self, company):
        self._companies.append(company)
    def advanced_list_companies(self):
        for i, company in enumerate(self._companies):
            print(f"->{i}|{company.name}")

        share_value_increase = float(0)

        if (len(company.share_value_history) >= 2):
            share_value_increase = round((1 - company.share_value_history[-2] / company.share_value), 3) * 100
        
        share_value_op = ""
        forecast_op = ["*", "*"]
        if share_value_increase >= 0:
            share_value_op = "+"
        else:
            share_value_op = "-"

        for n in range(2):
            if company.forecast[n] > 0:
                forecast_op[n] = "+"
            else:
                forecast_op[n] = "-"
        print(f"""
\tL share_value={round(company.share_value, 3)} ({round(share_value_increase, 3)}%)
\tL Forecast= {forecast_op[0]}{forecast_op[1]}""")

    def update(self):
        for i, company in enumerate(self._companies):
            forecast_roll = 0
            if random.randint(1, 100) >= 50 - company.forecast_influence * (company.forecast[0] + company.forecast[1] - 1):
                forecast_roll = 1
            else:
                forecast_roll = -1
            rate_of_change_roll = float(random.randint(0, round((company.rate_of_change[1] - company.rate_of_change[0])*1000))) / 1000 + company.rate_of_change[0]
            new_share_value = company.share_value + company.share_value * rate_of_change_roll * forecast_roll 
            
            company.share_value_history.append(new_share_value)
            company.share_value = company.share_value_history[-1]

            company.forecast = (random.randint(0, 1), random.randint(0, 1))

    def getCompanies(self):
        return self.companies

File: market.cpp
#include "market.h"
#include "myArith.h"
#include <iostream>

MyArith myArith;

Market::Market() {}

void Market::addCompany(Company& company) { companies.push_back(company); }
void Market::advancedListCompanies() {
    for (int i = 0; i < companies.size(); i++)
    {
        Company& company = companies[i];
        std::cout << "->" << i << "..." << company.getName() << "\n";

        float shareValueIncrease = 0;
        float maxChangeIncrease = 0;
        float minChangeIncrease = 0;

        if (company.getShareValueHistory().size() >= 2) {
            shareValueIncrease = myArith.roundToDecimal((1 - (company.getShareValueHistory()[company.getShareValueHistory().size() - 2] / company.getShareValue())) * 100, 3);
        }

        std::string shareValueOp = shareValueIncrease >= 0 ? "+" : "";
        std::string maxChangeOp = maxChangeIncrease >= 0 ? "+" : "";
        std::string minChangeOp = minChangeIncrease >= 0 ? "+" : "";
        std::pair<std::string, std::string> forecastOp = {
            company.getForecast().first == 1 ? "+" : "-",
            company.getForecast().second == 1 ? "+" : "-" };

        std::cout << "\tL ShareValue=" << company.getShareValue() << " ("
            << shareValueOp << shareValueIncrease << "%)\n";
        std::cout << "\tForecast=" << forecastOp.first << forecastOp.second
            << "\n";
    }
}

void Market::update()
{

    for (int i = 0; i < companies.size(); i++)
    {
        Company& company = companies[i];

        int forecastRoll = rand() % 100 >= 50 - company.getForecastInfluence() * (company.getForecast().first + company.getForecast().second - 1) ? 1 : -1;

        float rateOfChangeRoll = static_cast<float>(rand() % static_cast<int>((company.getRateOfChange().second - company.getRateOfChange().first) * 1000)) / 1000 + company.rateOfChange.first;
        float shareValueChange = company.getShareValue() + company.getShareValue() * forecastRoll * rateOfChangeRoll;

        company.getShareValueHistory().push_back(shareValueChange);
        company.shareValue = company.getShareValueHistory()[company.getShareValueHistory().size() - 1];
        std::pair<int, int> newForecast = { (rand() % 2), (rand() % 2) };
        company.forecast = newForecast;
    }
}

std::vector<Company> Market::getCompanies() { return companies; }